java/src/jmri/jmrix/can/adapters/gridconnect/GcSerialDriverAdapter.java

```diff
@@ -128,14 +133,155 @@ protected GcTrafficController makeGcTrafficController() {
+         * @return a block of data read, or nullptr if fatal IO errors make further use
+         * of this port impossible.
+         */
+        private BufferEntry tryRead(int len) {
+            BufferEntry tail = new BufferEntry();
+            try {
+                tail.data = new byte[len];
+                tail.len = in.read(tail.data, 0, len);
+                errorCount = 0;
+            } catch (IOException e) {
+                tail.e = e;
+                e.printStackTrace();
+                if (++errorCount > MAX_IO_ERRORS_TO_ABORT) {
+                    log.error("Closing read thread due to too many IO errors");
+                    return null;
+                }
+            }
+            return tail;
+        }
+
+        /**
+         * Implementation of the buffering thread.
+         */
+        private void readThreadBody() {
+            BufferEntry tail;
+            while(true) {
+                // Try to read one byte to block the thread.
+                tail = tryRead(1);
+                if (tail == null) return;
+                // NOTE: in order to reuse this class in a generic context, we need to add support
+                // for the underlying input stream persistently returning EOF. That does not
+                // happen on a serial port.
+                if (tail.len > 0 || tail.e != null) {
+                    readAhead.add(tail);
+                } else {
+                    continue;
+                }
+                // Read as many bytes as we have in large increments. REading 128 bytes is a good
+                // compromise between throughput (4 gridconnect packets per kernel IO) but not
+                // wasting a lot of memory if less data actually shows up.
+                do {
+                    tail = tryRead(128);
+                    if (tail == null) return;
+                    if (tail.len > 0 || tail.e != null) {
+                        readAhead.add(tail);
+                    } else {
+                        break;
+                    }
+                } while(true);
+            }
+        }
+
+        /**
+         * We queue objects of this class between the read thread and the actual read() methods.
+         */
+        private class BufferEntry {
+            // data payload
+            byte[] data;
+            // how many bytes of data are filled in
+            int len = 0;
+            // an exception was caught reading the input stream
+            IOException e = null;
+        }
+
+        @Override
+        public int read() throws IOException {
+            throw new IOException("unimplemented");
+        }
+
+        @Override
+        public int read(byte[] bytes) throws IOException {
+            throw new IOException("unimplemented");
+        }
+
+        @Override
+        public synchronized int read(byte[] bytes, int skip, int len) throws IOException {
+            if (skip != 0) {
+                throw new IOException("unimplemented");
```

Comment:
Throw an InvalidArgumentExecption if parameters are incorrect, not IOExceptions.

Reply:
I changed this to UnsupportedOPerationException as the API itself allows this value but we never use it so we don't need to support it here.